python/src/iceberg/avro/reader.py

```diff
@@ -0,0 +1,318 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""
+Classes for building the Reader tree
+
+Constructing a reader tree from the schema makes it easy
+to decouple the reader implementation from the schema.
+
+The reader tree can be changed in such a way that the
+read schema is different, while respecting the read schema
+"""
+from __future__ import annotations
+
+from abc import abstractmethod
+from dataclasses import dataclass, field
+from datetime import date, datetime, time
+from decimal import Decimal
+from functools import singledispatch
+from typing import Any
+from uuid import UUID
+
+from iceberg.avro.decoder import BinaryDecoder
+from iceberg.files import StructProtocol
+from iceberg.schema import Schema, SchemaVisitor
+from iceberg.types import (
+    BinaryType,
+    BooleanType,
+    DateType,
+    DecimalType,
+    DoubleType,
+    FixedType,
+    FloatType,
+    IntegerType,
+    ListType,
+    LongType,
+    MapType,
+    NestedField,
+    PrimitiveType,
+    StringType,
+    StructType,
+    TimestampType,
+    TimestamptzType,
+    TimeType,
+)
+from iceberg.utils.singleton import Singleton
+
+
+@dataclass(frozen=True)
+class AvroStruct(StructProtocol):
+    _data: list[Any | StructProtocol] = field()
+
+    def set(self, pos: int, value: Any) -> None:
+        self._data[pos] = value
```

Comment:
Where is `_data` initialized? If I create a list and try to assign to a position, it fails with `IndexError: list assignment index out of range`:

```python
In [1]: _data = []

In [2]: _data[3] = "foo"
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
Input In [2], in <cell line: 1>()
----> 1 _data[3] = "foo"

IndexError: list assignment index out of range
```

Reply:
Okay, I see that the data list is passed in. Seems okay for now.

Reply:
We can optimize this later by reusing objects. But I think it makes more sense to get everything working, and then do some profiling